upup/pkg/fi/cloudup/awstasks/natgateway.go

```diff
@@ -0,0 +1,151 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package awstasks
+
+import (
+	"fmt"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/golang/glog"
+	"k8s.io/kops/upup/pkg/fi"
+	"k8s.io/kops/upup/pkg/fi/cloudup/awsup"
+	"k8s.io/kops/upup/pkg/fi/cloudup/terraform"
+)
+
+//go:generate fitask -type=natgateway
+type NATGateway struct {
+	ID           *string // AWS ID for the object
+	AllocationID *string // Pointer to an elastic IP
+	SubnetID     *string // Pointer to a subnet in the VPC
+	VPCID        *string // Unique identifier of the VPC for this NAT gateway
+}
+
+var _ fi.CompareWithID = &NATGateway{} // Validate the IDs
+
+func (e *NATGateway) CompareWithID() *string {
+	return e.ID
+}
+
+func (e *NATGateway) Find(c *fi.Context) (*NATGateway, error) {
+	cloud := c.Cloud.(awsup.AWSCloud)
+
+	request := &ec2.DescribeNatGatewaysInput{}
+
+	if fi.StringValue(e.ID) != "" {
+		request.NatGatewayIds = []*string{e.ID}
+	} else {
+		request.Filter = cloud.BuildFilters(e.SubnetID)
+	}
+
+	response, err := cloud.EC2().DescribeNatGateways(request)
+	if err != nil {
+		return nil, fmt.Errorf("error listing NAT Gateways: %v", err)
+	}
+	if response == nil || len(response.NatGateways) == 0 {
+		return nil, nil
+	}
+
+	if len(response.NatGateways) != 1 {
+		return nil, fmt.Errorf("found multiple NAT Gateways matching tags")
+	}
+	ngw := response.NatGateways[0]
+	actual := &NATGateway{
+		ID:   ngw.NatGatewayId,
+		VPCID: ngw.VpcId,
+		SubnetID: ngw.SubnetId,
+	}
+
+	glog.V(4).Infof("found matching NAT gateway %v", actual)
+
+	// Allocation ID
+	if actual.ID != nil {
+		request := &ec2.DescribeAddressesInput{}
+		request.Filters = cloud.BuildFilters(e.VPCID)
+		response, err := cloud.EC2().DescribeAddresses(request)
+		if err != nil || len(response.Addresses) != 1 {
+			return nil, fmt.Errorf("error querying for elastic ip support: %v", err)
```

Comment:
I suspect we're going to want more info here so when that we have a way to diagnose when a user hits it

Also to differentiate the 0 vs >1 cases - again just for diagnostics.